File: solutions/2026-03-07/solution.java
// Minimum Window Substring
// Difficulty: Hard
// URL: https://leetcode.com/problems/minimum-window-substring/

public String minWindow(String s, String t) {
  if (s == null || t == null || s.length() == 0 || t.length() == 0) {
    return "";
  }
  int[] tCount = new int[128];
  int[] sCount = new int[128];
  for (char c : t.toCharArray()) {
    tCount[c]++;
  }
  int requiredChars = t.length();
  int left = 0;
  int minLen = Integer.MAX_VALUE;
  String minWindow = "";
  for (int right = 0; right < s.length(); right++) {
    if (tCount[s.charAt(right)] > 0) {
      sCount[s.charAt(right)]++;
      if (sCount[s.charAt(right)] == tCount[s.charAt(right)]) {
        requiredChars--;
      }
    }
    while (requiredChars == 0) {
      if (right - left + 1 < minLen) {
        minLen = right - left + 1;
        minWindow = s.substring(left, right + 1);
      }
      if (tCount[s.charAt(left)] > 0) {
        sCount[s.charAt(left)]--;
        if (sCount[s.charAt(left)] < tCount[s.charAt(left)]) {
          requiredChars++;
        }
      }
      left++;
    }
  }
  return minWindow;
}

File: solutions/2026-03-07/solution.py
# Minimum Window Substring
# Difficulty: Hard
# URL: https://leetcode.com/problems/minimum-window-substring/

def minWindow(s: str, t: str) -> str:
  if not s or not t:
    return ''
  t_count = {}
  s_count = {}
  for char in t:
    t_count[char] = t_count.get(char, 0) + 1
  required_chars = len(t_count)
  left = 0
  min_len = float('inf')
  min_window = ''
  for right in range(len(s)):
    if s[right] in t_count:
      s_count[s[right]] = s_count.get(s[right], 0) + 1
      if s_count[s[right]] == t_count[s[right]]:
        required_chars -= 1
    while required_chars == 0:
      if right - left + 1 < min_len:
        min_len = right - left + 1
        min_window = s[left:right + 1]
      if s[left] in t_count:
        s_count[s[left]] -= 1
        if s_count[s[left]] < t_count[s[left]]:
          required_chars += 1
      left += 1
  return min_window